cpp/src/arrow/compute/kernels/aggregate_mode.cc

```diff
@@ -205,35 +243,59 @@ struct ModeImpl : public ScalarAggregator {
     this->state.MergeFrom(std::move(other.state));
   }
 
-  void Finalize(KernelContext*, Datum* out) override {
-    using ModeType = typename TypeTraits<ArrowType>::ScalarType;
-    using CountType = typename TypeTraits<Int64Type>::ScalarType;
+  static std::shared_ptr<ArrayData> MakeArrayData(
+      const std::shared_ptr<DataType>& data_type, int64_t n) {
+    auto data = ArrayData::Make(data_type, n, 0);
+    data->buffers.resize(2);
+    data->buffers[0] = nullptr;
+    return data;
+  }
 
-    std::vector<std::shared_ptr<Scalar>> values;
-    auto mode_count = this->state.Finalize();
-    auto mode = mode_count.first;
-    auto count = mode_count.second;
-    if (count == 0) {
-      values = {std::make_shared<ModeType>(), std::make_shared<CountType>()};
-    } else {
-      values = {std::make_shared<ModeType>(mode), std::make_shared<CountType>(count)};
+  void Finalize(KernelContext* ctx, Datum* out) override {
+    const auto& mode_type = TypeTraits<ArrowType>::type_singleton();
+    const auto& count_type = int64();
+    const auto& out_type =
+        struct_({field(kModeFieldName, mode_type), field(kCountFieldName, count_type)});
+
+    int n = this->options.n;
+    if (n > state.DistinctValues()) {
+      n = state.DistinctValues();
+    }
+    if (n <= 0) {
+      *out = Datum(StructArray(out_type, 0, ArrayVector{}).data());
+      return;
     }
-    out->value = std::make_shared<StructScalar>(std::move(values), this->out_type);
+
+    auto mode_data = this->MakeArrayData(mode_type, n);
+    auto count_data = this->MakeArrayData(count_type, n);
+    KERNEL_ASSIGN_OR_RAISE(mode_data->buffers[1], ctx, ctx->Allocate(n * sizeof(CType)));
+    KERNEL_ASSIGN_OR_RAISE(count_data->buffers[1], ctx,
+                           ctx->Allocate(n * sizeof(int64_t)));
+
+    CType* mode_buffer = mode_data->template GetMutableValues<CType>(1);
+    int64_t* count_buffer = count_data->template GetMutableValues<int64_t>(1);
+    this->state.Finalize(mode_buffer, count_buffer, n);
+
+    *out = Datum(
+        StructArray(out_type, n, ArrayVector{MakeArray(mode_data), MakeArray(count_data)})
```

Comment:
You could perhaps construct the `ArrayData` directly instead of going through a temporary `StructArray`.

Reply:
Done